### DISCUSSION TITLE
Firefly III generates insecure URLs when behind a secure reverse proxy

### DISCUSSION THREAD

**Discussion by searing** [firefly-iii/firefly-iii#9190] [Firefly III]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

I'm trying to setup Firefly III to run in Docker behind an nginx instance running on the host.

This is the intended flow:

web browser <--- port 443, TLS ---> nginx <--- port 8051 on host, 8080 in container ---> Firefly III

Here's the pertinent info from the systemd service I created. NOTE: throughout I will use "https://my.url.com" as the intended external URL. This obviously isn't the actual URL I'm using. I don't wish to share that publicly, but it's a standard .com domain and I have a TLS certificate from LetsEncrypt. The DNS and TLS settings appear to work fine so I think are outside the scope of this issue.

```
ExecStart=/usr/bin/docker run --rm --name fireflyiii \
            -p 8051:8080 \
            -v /storage/fireflyiii/upload:/var/www/html/storage/upload \
            -e APP_KEY={{ fireflyiii_api_key }} \
            -e DB_HOST=(redacted, db host) \
            -e DB_PORT=5432 \
            -e DB_CONNECTION=pgsql \
            -e DB_DATABASE=fireflyiii \
            -e DB_USERNAME=fireflyiii \
            -e DB_PASSWORD={{ fireflyiii_db_password }} \
            -e TZ=America/Chicago \
            -e APP_URL=https://my.url.com \
            fireflyiii/core:version-{{ fireflyiii_version }}
```

I can browse to http://my.url.com:8051 and access the site over regular HTTP. (Eventually I will try to disable this so all requests must go through nginx, but it's useful for testing.) However, if I go to https://my.url.com, I see a website with no styles or javascript. Looking at the firefox console, I see that the website is trying to load stuff from http://my.url.com.

For example:
> Blocked loading mixed active content “http://my.url.com/build/assets/dashboard-BZG_lvWP.js”

It *should* be loading it from "https://my.url.com/build/assets/dashboard-BZG_lvWP.js" but it's not. Presumably because the proxy connection is regular HTTP, the web service assumes all generated URLs should use that scheme as well, but that's not the case.

This could be a case of simply missing a configuration option, although I looked through the entire example .env file and didn't see anything like an external URL setting. The closest was APP_URL but I think that is only used in some contexts.

Is there a way for me to get Firefly III to generate URLs (ie for page assets, javascript + styles) using a specific protocol? Maybe a header I can set from nginx?

For completeness, here's the relevant nginx config. The nginx TLS config seems to work fine, and the right certificate is served when browsing there.
```
    server {
        listen 443 ssl;
        server_name my.url.com;

        ssl_certificate /etc/letsencrypt/live/my.url.com/fullchain.pem;
        ssl_certificate_key /etc/letsencrypt/live/my.url.com/privkey.pem;
        
        location / {
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_pass http://1.2.3.4:8051; # (note: my config file has the real IP here)
        }
    }
```

### Debug information

Debug information generated at 2024-09-01 11:11:21 for Firefly III version **v6.1.19**.

<table><thead><tr><th colspan="2">System information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>Firefly III</td><td>6.1.19 / v2.1.0 / 24 (exp. 24)</td></tr><tr><td>PHP version</td><td>8.3.9 (64bits) / apache2handler / Linux x86_64</td></tr><tr><td>BCscale</td><td>12</td></tr><tr><td>Error reporting</td><td>Display: Off, reporting: ALL errors</td></tr><tr><td>Max upload</td><td>67108864 (64 MB)</td></tr><tr><td>Database drivers</td><td>mysql, <strong>*pgsql*</strong>, sqlite, </td></tr><tr><td>Docker build</td><td><span>#</span>1113, base <span>#</span>86</td></tr></tbody></table><table><thead><tr><th colspan="2">Firefly III information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>Timezone</td><td>America/Chicago + America/Chicago</td></tr><tr><td>App environment</td><td>production, debug: false</td></tr><tr><td>Layout</td><td>v1</td></tr><tr><td>Logging</td><td>info, stack / (empty)</td></tr><tr><td>Cache driver</td><td>file</td></tr><tr><td>Default language and locale</td><td>en_US + equal</td></tr><tr><td>Trusted proxies</td><td></td></tr><tr><td>Login provider & user guard</td><td>eloquent / web</td></tr><tr><td>Login headers</td><td>N/A + N/A</td></tr><tr><td>Stateful domains</td><td></td></tr><tr><td>Last cron job</td><td>never (never)</td></tr><tr><td>Mailer</td><td>log</td></tr></tbody></table><table><thead><tr><th colspan="2">User-specific information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>User</td><td><span>#</span>1 of 1</td></tr><tr><td>User flags</td><td></td></tr><tr><td>Session start</td><td>2024-09-01 00:00:00</td></tr><tr><td>Session end</td><td>2024-09-30 23:59:59</td></tr><tr><td>View range</td><td>1M</td></tr><tr><td>User language</td><td>en_US</td></tr><tr><td>User locale</td><td>en_US</td></tr><tr><td>Locale(s) supported</td><td>en_US.utf8: :white_check_mark:<br>en_US.UTF-8: :white_check_mark:<br></td></tr><tr><td>User agent</td><td>Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:128.0) Gecko/20100101 Firefox/128.0</td></tr></tbody></table>

**Comment by JC5** [comment]
Thank you for opening an issue. Since this is a configuration problem on your end, I've converted it to a Q&A discussion.

Please set `TRUSTED_PROXIES=*` or `TRUSTED_PROXIES=(IP of your reverse proxy)`. Feel free to consult the documentation if you have questions.

**Reply by searing** [reply]
Ah, sorry for mis-categorizing it.

I tried setting `TRUSTED_PROXIES` to `*` and `**` via the `docker` command (ie. `-e TRUSTED_PROXIES=*`), but that didn't work.

However, after searching the issues here for "TRUSTED_PROXIES" I found some others having the same issue and found a solution. I simply had to set the X-Forwarded-Proto header in nginx, in addition to the other headers I was already setting.

```
proxy_set_header X-Forwarded-Proto $scheme;
```

Now Firefly III generates https links for page content.

Thanks for the quick response!
